Tests/SourceKitLSPTests/SemanticTokensTests.swift

```diff
@@ -0,0 +1,252 @@
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2021 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import LanguageServerProtocol
+import LSPTestSupport
+import SKTestSupport
+import SourceKitLSP
+import XCTest
+
+final class SemanticTokensTests: XCTestCase {
+  /// Connection and lifetime management for the service.
+  var connection: TestSourceKitServer! = nil
+
+  /// The primary interface to make requests to the SourceKitServer.
+  var sk: TestClient! = nil
+
+  override func tearDown() {
+    sk = nil
+    connection = nil
+  }
+
+  override func setUp() {
+    connection = TestSourceKitServer()
+    sk = connection.client
+    _ = try! sk.sendSync(InitializeRequest(
+      processId: nil,
+      rootPath: nil,
+      rootURI: nil,
+      initializationOptions: nil,
+      capabilities: ClientCapabilities(
+        workspace: .init(
+          semanticTokens: .init(
+            refreshSupport: true
+          )
+        ),
+        textDocument: nil
+      ),
+      trace: .off,
+      workspaceFolders: nil
+    ))
+  }
+
+  private func performSemanticTokensRequest(text: String, range: Range<Position>? = nil) -> [SemanticToken] {
+    let url = URL(fileURLWithPath: "/\(#function)/a.swift")
+
+    // We wait for the first refresh request to make sure that the semantic tokens are ready
+
+    let expectation = XCTestExpectation(description: "performSemanticTokensRequest - refresh received")
+    sk.handleNextRequest { (req: Request<WorkspaceSemanticTokensRefreshRequest>) in
+      expectation.fulfill()
+      req.reply(VoidResponse())
+    }
+
+    sk.send(DidOpenTextDocumentNotification(textDocument: TextDocumentItem(
+      uri: DocumentURI(url),
+      language: .swift,
+      version: 17,
+      text: text
+    )))
+
+    let waitResult = XCTWaiter.wait(for: [expectation], timeout: 15)
+    if waitResult != .completed {
+      fatalError("error \(waitResult) while waiting for semantic token refresh")
+    }
+
+    let textDocument = TextDocumentIdentifier(url)
+    let response: DocumentSemanticTokensResponse!
+
+    if let range = range {
+      response = try! sk.sendSync(DocumentSemanticTokensRangeRequest(textDocument: textDocument, range: range))
+    } else {
+      response = try! sk.sendSync(DocumentSemanticTokensRequest(textDocument: textDocument))
+    }
+
+    return decodeFromIntArray(rawSemanticTokens: response.data)
+  }
+
+  func testEmpty() {
+    let text = ""
+    let tokens = performSemanticTokensRequest(text: text)
+    XCTAssertEqual(tokens, [])
+  }
+
+  func testRanged() {
+    let text = """
+    let x = 1
+    let y = 2
+    let z = 3
+    let w = 4
```

Comment:
Would it be worth to have variables/numbers of different lengths? At the moment, the test case would also succeed if we returned results for lines 0 and 1 instead of 1 and 2.